db/replication_epoch_edit.cc

```diff
@@ -0,0 +1,142 @@
+Status ReplicationEpochAddition::DecodeFrom(Slice* input) {
+    if (!GetVarint64(input, &epoch_)) {
+        return Status::Corruption("ReplicationEpochAddition", "Error decoding epoch");
+    }
+    if (!GetVarint64(input, &first_mus_)) {
+        return Status::Corruption("ReplicationEpochAddition", "Error decoding manifest update sequence");
+    }
+    return Status::OK();
+}
+
+std::string ReplicationEpochAddition::DebugString() const {
+    std::ostringstream oss;
+    oss << *this;
+    return oss.str();
+}
+
+std::ostream& operator<<(std::ostream& os, const ReplicationEpochAddition& ea) {
+    os << "epoch: " << ea.GetEpoch() << ", mus: " << ea.GetFirstMUS();
+    return os;
+}
+
+Status ReplicationEpochSet::AddEpoch(const ReplicationEpochAddition& epoch,
+                                     uint32_t max_num_replication_epochs) {
+  if (!epochs_.empty() && epochs_.back() >= epoch) {
+    std::stringstream ss;
+    ss << "Misordered replication epoch. prev: " << epochs_.back()
+       << ", next: " << epoch;
+    return Status::Corruption(ss.str());
+  }
+
+  while (epochs_.size() + 1 >= max_num_replication_epochs) {
+    epochs_.pop_front();
+  }
+  epochs_.push_back(epoch);
+  return Status::OK();
+}
+
+Status ReplicationEpochSet::AddEpochs(const ReplicationEpochAdditions& epochs, uint32_t max_num_replication_epochs) {
+    for (auto& epoch: epochs) {
+      auto s = AddEpoch(epoch, max_num_replication_epochs);
+      if (!s.ok()) {
+        return s;
+        }
+    }
+    return Status::OK();
+}
+
+
+Status ReplicationEpochSet::VerifyNewEpochs(const ReplicationEpochAdditions& new_epochs) const {
+    if (new_epochs.empty()) {
+        return Status::OK();
+    }
+
+    std::optional<ReplicationEpochAddition> prev;
+    if (!epochs_.empty()) {
+      prev = epochs_.back();
+    }
+
+    for (auto& epoch: new_epochs) {
+        if (!prev) {
+            prev = epoch;
+            continue;
+        }
+
+        if (*prev >= epoch) {
+          std::stringstream ss;
+          ss << "Misordered replication epoch. prev: " << *prev
+             << ", next: " << epoch;
+          return Status::Corruption(ss.str());
+        }
+
+        prev = epoch;
+    }
+    return Status::OK();
+}
+
+void ReplicationEpochSet::DeleteEpochsBefore(uint64_t epoch) {
+  while (!epochs_.empty() && (epochs_.front().GetEpoch() < epoch)) {
+    epochs_.pop_front();
+  }
+}
+
+std::optional<uint64_t> ReplicationEpochSet::GetEpochForMUS(uint64_t mus) const {
+    if (empty() || epochs_.front().GetFirstMUS() > mus) {
+        return std::nullopt;
+    }
+
+    // fast path
+    if (epochs_.back().GetFirstMUS() <= mus) {
+        return epochs_.back().GetEpoch();
+    }
+
+    auto it = std::upper_bound(epochs_.begin(), epochs_.end(), mus,
+                               [&](auto mus, const auto& epoch_addition) {
+                                 return mus < epoch_addition.GetFirstMUS();
+                               });
+    assert(it != epochs_.begin());
```

Comment:
It seems a bit safer to write this function as:

```
    if (!epochs_.empty() && epochs_.back().GetFirstMUS() <= mus) {
        return epochs_.back().GetEpoch();
    }
    
        auto it = std::upper_bound(epochs_.begin(), epochs_.end(), mus,
                               [&](auto mus, const auto& epoch_addition) {
                                 return mus < epoch_addition.GetFirstMUS();
                               });
       if (it == epochs_.begin()) { return std::nullopt; }
       
       --it;
       ...
```

That way you don't need the assert (which is compiled out in debug mode) and are certain that `--it` is safe.